Fixes for ublox SARA modem driver

mike-scott opened the pull request:
This PR is a combination of fixes for a wide variety of issues noticed on the SARA-R4 modem:
- MQTT wasn't working correctly
- http_get sample was crashing
- echo_client wasn't working
- odd behavior was happening in LwM2M client under load

Some additional authors contributed patches which were added here for easy testing due to the # of changes:
@weinholtendian and @hwilmers

Fixes: https://github.com/zephyrproject-rtos/zephyr/issues/21764
Fixes: https://github.com/zephyrproject-rtos/zephyr/issues/21328

Commit:
drivers: modem: sara-r4: continually signal the modem to power off

The U2 modem will sometimes not power off even if we have given it the
signal to do so. We now signal it continually until the modem
indicates that power is off with VINT=0.

Signed-off-by: Göran Weinholt <goran.weinholt@endian.se>

Commit:
drivers: modem: sara-r4: support for network indication on SARA modems

The u-blox SARA modems have the ability to output the network status
on a GPIO pin. This can be used to light up a LED when the modem is
connected to the network.

Signed-off-by: Göran Weinholt <goran.weinholt@endian.se>

Commit:
drivers: modem: sara-r4: support for 2G on the SARA R412M modem

The SARA R412M requires manual activation of the PDP context. It also
reports RSSI instead of RSRP when on 2G. An off-by-one in the RSRP
calculation was also fixed in this commit.

Signed-off-by: Göran Weinholt <goran.weinholt@endian.se>

mike-scott committed:
drivers: modem: sara-r4: return send bytes from send_socket_data

The send_socket_data() is incorrectly sending 0 as successful result.

The socket APIs require the number of bytes sent to be returned.

Signed-off-by: Michael Scott <mike@foundries.io>

mike-scott committed:
drivers: modem: sara-r4: space for NUL in buffer is optional

The hex_to_binary() function is incorrectly assuming all buffers
passed to recv() will have an additinal byte for storing a NUL
terminating char.  This should be optional as MQTT library uses
exact sized buffers for parsing socket data.

Signed-off-by: Michael Scott <mike@foundries.io>

mike-scott committed:
drivers: modem: sara-r4: dont unblock on sockread

Users may be waiting for data via socket recv() function.  We
notify them that data is ready when URC is received from modem.

Once unblocked, we read the data from the modem buffer which
is handled via on_cmd_sockread_common().  At this point, we
don't need to unblock waiting users again.

Signed-off-by: Michael Scott <mike@foundries.io>

mike-scott committed:
drivers: modem: sara-r4: cleanup offload_recvfrom timeout

In offload_recvfrom(), if we end up calling modem_cmd_send(),
then we should always wait for the results.  We've already
checked for MSG_DONTWAIT and made sure that there is data ready.
We just need to get it from the modem buffer at this point.

Signed-off-by: Michael Scott <mike@foundries.io>

mike-scott commented:
These fixes are related to: https://github.com/zephyrproject-rtos/zephyr/issues/21764  But don't quite fix all issues.  MQTT is still having problems due to the way it uses recv() with perfectly sized buffers to parse the socket data.

The SARA-R4 modem AT commands may have a bug where the full packet data needs to be queried at a time.

mike-scott commented:
@gpaquet85 @weinholtendian @WilliamGFish FYI

mike-scott committed:
drivers: modem: sara-r4: remove interface up / down

Managing the interface up / down events from driver code (as opposed
to L2 layers) has been an issue for quite some time.  There are
race conditions which result in data aborts (referencing NULL
iface or NULL iface->if_dev) or ends up breaking ASSERT checks for
non-NULL iface.

Let's remove the handling for now and come up with something better
when a user actually needs it.

Signed-off-by: Michael Scott <mike@foundries.io>

brooksprumo reviewed (approved):
I was previously running into an assertion fault because a network interface pointer was NULL and being dereferenced, when using Particle Boron. This PR fixed that crash. Yay!

mike-scott commented:
NOTE: More work coming on this PR.

jukkar commented:
> MQTT is still having problems due to the way it uses recv() with perfectly sized buffers to parse the socket data.

@mike-scott What kind of problems are there in MQTT, can you elaborate more?
cc: @rlubos

mike-scott commented:
> > MQTT is still having problems due to the way it uses recv() with perfectly sized buffers to parse the socket data.
> 
> @mike-scott What kind of problems are there in MQTT, can you elaborate more?
> cc: @rlubos

@jukkar @rlubos 
Yes, sorry to be a bit vague.  I was trying to make better use of the modem's data buffer by only pulling data as it was needed via the recv() call (based on the length of the buffer passed in).

The MQTT library parses individual packets bytes at a time, so this leads to successive calls to the modem to retrieve data for the same packet.  The way this modem driver works is it keeps track of how much data in in the modem's buffer (split up by packet). My handling of the packet sizes didn't allow for partial retrieval of packets in this way.

I need to think about how to handle this.

hwilmers committed:
drivers: modem: sara-r4: adjust send and receive lengths

- limit max. number of bytes when sending to socket
  The number of bytes sent to a socket in one transaction
  is limited to 512 in HEX mode (Sara-R4), and to 1024
  otherwise. This corresponds to numbers given in the
  manual for ublox cellular modems.

- report number of bytes actually sent, as reported by modem
  After writing data to a socket, we now return the number of
  bytes actually written, as reported by the modem.

Signed-off-by: Hans Wilmers <hans@wilmers.no>

hwilmers committed:
drivers: modem: sara-r4: fix modem reset

This fixes two problems in modem_reset():

- mdata.net_iface is used without checking that it is valid.
  Now, we validate mdata_net_iface before usage, and give a warning
  if it was not set.

- if the modem does not perform network registration within the
  given time of 20s, in the previous implementation the driver
  gave up.
  Now, the driver disables RF for a second, enables it again and
  gives the modem another 20s to register. This has been seen
  to help in roaming situations.

Tested on Sara R4.

Signed-off-by: Hans Wilmers <hans@wilmers.no>

hwilmers commented:
Nice fixes, this works here with a Sara R410.

mike-scott committed:
drivers: modem: socket: add wait_data and data_ready functions

Let's hide the internals of the modem_socket's sem_data_ready and
poll handling with 2 new functions:
- modem_socket_wait_data: take a semaphore and wait for data
- modem_socket_data_ready: give back the data ready semaphore and
  unblock poll() users

Signed-off-by: Michael Scott <mike@foundries.io>

mike-scott committed:
drivers: modem: socket: allow partial packet reads

Let's allow protocols to request portions of the current packet.

This is seen in the MQTT library when parsing headers for the type
and variable length of the packet.

This fixes basic parsing done in the MQTT library and probably others.

Signed-off-by: Michael Scott <mike@foundries.io>

mike-scott committed:
drivers: modem: socket: introduce locking

Add lock behavior for functions in modem_socket, to prevent race
conditions when performing socket data maintenance.

Signed-off-by: Michael Scott <mike@foundries.io>

mike-scott committed:
drivers: modem: socket: reset is_polled in modem_socket_put()

Let's make sure is_polled gets reset when we reset the socket.

Signed-off-by: Michael Scott <mike@foundries.io>

mike-scott committed:
drivers: modem: socket: add modem_socket_next_packet_size

Let's hide the internals of sock->packet_sizes[] by adding a function
which returns the size of the next waiting packet.

Signed-off-by: Michael Scott <mike@foundries.io>

mike-scott committed:
drivers: modem: socket: add is_connected field to modem_socket

Let's add a field which the drivers can use to keep track of whether
they are connected or not.  This will normally be enabled / disabled
in the socket connect and URC for socket close notify.

Signed-off-by: Michael Scott <mike@foundries.io>

mike-scott committed:
drivers: modem: cmd_handler: cleanup parse_params()

- rename buf_len to match_len for clarity
- pass in modem_cmd_handler_data instead of buffer pointer directly
- buffer adjustments for the command length are kept inside
  parse_params()

Signed-off-by: Michael Scott <mike@foundries.io>

mike-scott committed:
drivers: modem: cmd_handler: save 1 char for NUL in match_buf

For parsing purposes we need to add a NUL to the end of match_buf.

When there is no CR/LF in the incoming rx bufs then we fill match_buf
to its max size.  This ended up with an off by one error which was
overflowing match_buf into the following data.

To account for this, let's fill the buffer to size - 1 so that we
leave room for the NUL at the end and stop corrupting data.

Signed-off-by: Michael Scott <mike@foundries.io>

mike-scott committed:
drivers: modem: sara-r4: make room for NUL in match_buf

In order to still process 128 bytes at a time, let's add 1 to the
length of match_buf which will be used to store a NUL char when
the match_buf is at max size (size - 1).

Otherwise we're processing 127 bytes at a time which is inefficient
compared to the data buffer sizes in net_buf.

Signed-off-by: Michael Scott <mike@foundries.io>

mike-scott committed:
boards/shields: sparkfun_sara_r4: enable SARA_R4_NET_STATUS

There is a blue LED used to display connectivity for the SparkFun
SARA-R4 shield.  This LED is connected to the modem via GPIO 16
which is the default used by SARA-R4 driver when
CONFIG_MODEM_UBLOX_SARA_R4_NET_STATUS is enabled.

Signed-off-by: Michael Scott <mike@foundries.io>

mike-scott committed:
drivers: modem: sara-r4: use 120s connection timeout

SARA-R4 AT command manual states +USOCO has a max timeout of 120s.
Let's use that instead of the default 10s command timeout.

Signed-off-by: Michael Scott <mike@foundries.io>

mike-scott committed:
drivers: modem: sara-r4: manage sock->is_connected

Let's use sock->is_connected to check whether we need to clean up
the modem's socket.  It may have already been closed via URC.

Signed-off-by: Michael Scott <mike@foundries.io>

mike-scott committed:
drivers: modem: sara-r4: use new socket functions

New socket functions were introduced to hide some of the modem_socket
internals:
- modem_socket_next_packet_size()
- modem_socket_wait_data()
- modem_socket_data_ready()

Let's use them.

Signed-off-by: Michael Scott <mike@foundries.io>

mike-scott committed:
drivers: modem: sara-r4: fix error text in net_offload_dummy_get

Correctly tell the user that CONFIG_NET_SOCKETS_OFFLOAD needs to be
enabled.

Signed-off-by: Michael Scott <mike@foundries.io>

mike-scott committed:
drivers: modem: sara-r4: handle read limits

SARA-R4 has a max read limit of 512 when using binary hex formatting.
Otherwise the max read limit is 1024.

Signed-off-by: Michael Scott <mike@foundries.io>

mike-scott commented:
@gpaquet85 @WilliamGFish @weinholtendian @hwilmers
Can you please retest with the full set of patches?

NOTE: To test MQTT, you will also need this PR: https://github.com/zephyrproject-rtos/zephyr/pull/22249

mike-scott commented:
Here is the way I tested 3 of the samples for the SparkFun SARA-R4 shield (on a K64F):

LwM2M client (connecting to the IP of http://leshan.eclipse.org/):
```
west build -b frdm_k64f -s zephyr/samples/net/lwm2m_client -- \
  -DSHIELD=sparkfun_sara_r4 \
  -DCONFIG_NET_CONFIG_PEER_IPV4_ADDR=\"23.97.187.154\" \
  -DCONFIG_NET_IPV6=n -DCONFIG_NET_CONFIG_NEED_IPV6=n \
  -DCONFIG_MODEM_UBLOX_SARA_R4_APN=\"<APN>\" \
  -DCONFIG_MODEM_UBLOX_SARA_R4_MANUAL_MCCMNO=\"<manual MCCMNO>\" \
  -DCONFIG_COAP_INIT_ACK_TIMEOUT_MS=10000 -DCONFIG_MODEM_SHELL=y \
  -DCONFIG_LWM2M_SERVER_DEFAULT_PMIN=1 \
  -DCONFIG_LWM2M_SERVER_DEFAULT_PMAX=360 \
  -DCONFIG_LOG_STRDUP_MAX_STRING=128 \
  -DCONFIG_LOG_STRDUP_BUF_COUNT=10
```

HTTP GET sample:
```
west build -b frdm_k64f -s zephyr/samples/net/sockets/http_get -- \
  -DSHIELD=sparkfun_sara_r4 \
  -DCONFIG_MODEM_UBLOX_SARA_R4_APN=\"<APN>\" \
  -DCONFIG_MODEM_UBLOX_SARA_R4_MANUAL_MCCMNO=\"<manual MCCMNO>\" \
  -DCONFIG_MODEM_SHELL=y \
  -DCONFIG_LOG_IMMEDIATE=n \
  -DCONFIG_NET_CONFIG_NEED_IPV4=y
```

MQTT Client (also need the changes here: https://github.com/zephyrproject-rtos/zephyr/pull/22249)
```
west build -b frdm_k64f -s zephyr/samples/net/mqtt_publisher -- \
  -DSHIELD=sparkfun_sara_r4 \
  -DCONFIG_NET_CONFIG_PEER_IPV4_ADDR=\"5.196.95.208\" \
  -DCONFIG_NET_IPV6=n -DCONFIG_NET_CONFIG_NEED_IPV6=n \
  -DCONFIG_MODEM_UBLOX_SARA_R4_APN=\"<APN>\" \
  -DCONFIG_MODEM_UBLOX_SARA_R4_MANUAL_MCCMNO=\"<manual MCCMNO>\" \
  -DCONFIG_MODEM_SHELL=y \
  -DCONFIG_NET_CONFIG_NEED_IPV4=y
```

WilliamGFish commented:
> Here is the way I tested 3 of the samples for the SparkFun SARA-R4 shield (on a K64F):
> 
Now I see why you have been avoiding NET_OFFLOAD, I am using the Particle boards which use the U2, R4 and esp32 none of which have network port. This is why I have been suggesting to continue with the network offloading standard.

Maybe the nice guys at Nordic will send you a couple of Dev Kits to experiment with as the SparkFun shields should be compatible.

mike-scott commented:
> Now I see why you have been avoiding NET_OFFLOAD, I am using the Particle boards which use the U2, R4 and esp32 none of which have network port. This is why I have been suggesting to continue with the network offloading standard.
> 
> Maybe the nice guys at Nordic will send you a couple of Dev Kits to experiment with as the SparkFun shields should be compatible.

I have all 3 Particle.IO boards: Xenon, Argon and Boron.  I don't have a VZW SIM for use on my Boron so it makes things difficult, but I can certainly help with the Argon ESP8622 bring up (I'm looking at the PR: https://github.com/zephyrproject-rtos/zephyr/pull/21492)

Other than TLS support and possibly DNS, what functionality were you looking for on the NET_OFFLOAD APIs?

Looking at esp8622, it has internal methods for handling SSL-based connections which would need to use the setsockopt offloaded APIs (AT+CIPSSLCONF).

WilliamGFish commented:
> I have all 3 Particle.IO boards: Xenon, Argon and Boron. I don't have a VZW SIM for use on my Boron so it makes things difficult, but I can certainly help with the Argon ESP8622 bring up (I'm looking at the PR: #21492)

Be warned you will need to change the firmware on the esp32 before you can use it as the particle firmware is "locked down" so will give you odd results.

mike-scott commented:
> Be warned you will need to change the firmware on the esp32 before you can use it as the particle firmware is "locked down" so will give you odd results.

Was there a specific version to use from this list?
https://www.espressif.com/en/support/download/at

WilliamGFish commented:
> > Be warned you will need to change the firmware on the esp32 before you can use it as the particle firmware is "locked down" so will give you odd results.
> 
> Was there a specific version to use from this list?
> https://www.espressif.com/en/support/download/at

You may find this useful: https://learn.adafruit.com/adding-a-wifi-co-processor-to-circuitpython-esp8266-esp32/program-particle-argon

tsvehagen commented:
> > Now I see why you have been avoiding NET_OFFLOAD, I am using the Particle boards which use the U2, R4 and esp32 none of which have network port. This is why I have been suggesting to continue with the network offloading standard.
> > Maybe the nice guys at Nordic will send you a couple of Dev Kits to experiment with as the SparkFun shields should be compatible.
> 
> I have all 3 Particle.IO boards: Xenon, Argon and Boron. I don't have a VZW SIM for use on my Boron so it makes things difficult, but I can certainly help with the Argon ESP8622 bring up (I'm looking at the PR: #21492)
> 
> Other than TLS support and possibly DNS, what functionality were you looking for on the NET_OFFLOAD APIs?
> 
> Looking at esp8622, it has internal methods for handling SSL-based connections which would need to use the setsockopt offloaded APIs (AT+CIPSSLCONF).

The main reason for me _not_ wanting to offload SSL in the esp8266, or any external device really, is that data will be passed in clear text over the UART.

WilliamGFish commented:
> The main reason for me _not_ wanting to offload SSL in the esp8266, or any external device really, is that data will be passed in clear text over the UART.

I too am using a plain text buffer to feed the offloaded network i/f via the UART

hwilmers commented:
I tested this PR with Sara R410, with following results:
- MQTT client works fine
- echo_client (TCP) works fine with small chunks (up to 1023 bytes)
- echo_client (TCP) fails with chunks >= 1024 bytes

I then discovered that in case of these failures, the +USOWR fails with an error message from the modem ("ERROR"), immediately followed by a +UUSORD message from the modem.

To me it looks like a solution could be to let the application poll for incoming data in between multiple calls to send(). So should this be solved in the driver, or rather in the application?

hwilmers commented:
Apropos TLS: Is there a way to use this driver with TLS, other than offloading it? Are there any hooks one can use to use mbedTLS?

WilliamGFish commented:
> Apropos TLS: Is there a way to use this driver with TLS, other than offloading it? Are there any hooks one can use to use mbedTLS?

There are options to use mbedTLS in zephyr (CONFIG_MBEDTLS_BUILTIN)...it might but something to look at.

mike-scott commented:
> I tested this PR with Sara R410, with following results:
> 
> * MQTT client works fine
> * echo_client (TCP) works fine with small chunks (up to 1023 bytes)
> * echo_client (TCP) fails with chunks >= 1024 bytes
> 
> I then discovered that in case of these failures, the +USOWR fails with an error message from the modem ("ERROR"), immediately followed by a +UUSORD message from the modem.
> 
> To me it looks like a solution could be to let the application poll for incoming data in between multiple calls to send(). So should this be solved in the driver, or rather in the application?

Thanks for testing, let me take a look at the >= 1024 use case to see where the fix would best be located.  In a perfect world, we fix that in the driver so that samples don't need to be changed.  However, there is some "bad behavior" in the samples sometimes which can be changed.

mike-scott commented:
> The main reason for me _not_ wanting to offload SSL in the esp8266, or any external device really, is that data will be passed in clear text over the UART.

When I looked at how the SSL certificates would be setup for ESP8266, it seems like they needed to be written to flash to avoid this situation.

`AT+CIPSSLCCONF=2` means (taken from AT command doc):
```
bit1:if set to be 1, CA will be enabled, so ESP8266 can verify SSL server; if 0, then will not.
```

And the example AT command uses are:
```
AT+CIPSNTPCFG=1,8 // set SNTP timezone
AT+CIPSNTPTIME? // get SNTP time
AT+CIPSSLCCONF=2
AT+CIPSTART="SSL","192.168.3.38",8443
```

There doesn't seem to be a time when cleartext SSL information would be used over UART.

tsvehagen commented:
> There doesn't seem to be a time when cleartext SSL information would be used over UART.

Right but then when you send or read data over that connection the data will be sent unencrypted over the UART? Maybe I'm missing something 🤔

mike-scott commented:
> Right but then when you send or read data over that connection the data will be sent unencrypted over the UART? Maybe I'm missing something

Yes, this would be true.  UART lines are a fairly easy physical attack vector.
Unfortunately, this isn't limited to SSL connections.  The Wifi AP password is sent in clear text, etc.

mike-scott commented:
Updated with 2 patches to fixup the return values for socket offload and vtable functions.  They weren't standard POSIX values.

gpaquet85 commented:
@mike-scott really thanks for your support on this case.
I will made some tries next week and let you know.
I had also a look on MQTT PR. I will integrate it to know if it is ok from my side.

WilliamGFish commented:
> Updated with 2 patches to fixup the return values for socket offload and vtable functions. They weren't standard POSIX values.

Some sterling work @mike-scott , I'll pull this PR and do some testing at the beginning of next week.

mike-scott commented:
I have a few more patches to get on top of this PR, but it needs to be rebased now with the socket offload rework being merged.  The TOPIC-GPIO branch will also cause conflicts once it's merged.

mike-scott committed:
drivers: modem: sara-r4: add offloaded DNS handling

SARA-R4 modem supports offloaded DNS via AT+UDNSRN command.
Let's implement it.

NOTE! On SARA-R4 a new firmware *IS* required to support this feature:
L0.0.00.00.05.08 [Apr 17 2019 19:34:02]

Signed-off-by: Michael Scott <mike@foundries.io>

mike-scott committed:
drivers: modem: sara-r4: update functions to return POSIX values

The original Ublox SARA-R4 driver was written for the net_context
APIs.  As a result many of the return values are not POSIX standard.

Let's go through all of the socket offload functions and make sure
we return standard values (0 or -1) and set errno where appropriate.

Signed-off-by: Michael Scott <mike@foundries.io>

mike-scott committed:
drivers: modem: cmd_handler: remove extra comment

Remove extra comment left over from prototype driver.

Signed-off-by: Michael Scott <mike@foundries.io>

gpaquet85 commented:
@mike-scott I made some tries. For the moment basic MQTT publish is ok. Thanks for this. However I still have some fails when I run MQTT publish loop. I investigate right now on this and let you know.

mike-scott committed:
drivers: modem: cmd_handler: fix findcrlf() indentation

Missed during previous rework.

Signed-off-by: Michael Scott <mike@foundries.io>

mike-scott committed:
drivers: modem: cmd_handler: honor num of params in parse_params

Modem commands are setup with arg_count to denote the number of params
to parse before returning.  Let's honor that setting and return once
we've parse them.

This fixes a run-on bug where the parser would find as many of the
parameters via the supplied delimiter as it could.

Signed-off-by: Michael Scott <mike@foundries.io>

mike-scott committed:
drivers: modem: cmd_handler: add return value to modem commands

Some modem commands can determine if they have enough data pulled from
the modem to continue or not.  Let's allow those functions to return
EAGAIN which means there's more data needed from the modem.

Signed-off-by: Michael Scott <mike@foundries.io>

mike-scott committed:
drivers: modem: sara-r4: disable hex mode for binary data

Now that the u-blox driver can let cmd_handler know that it hasn't
received enough data, let's disable the hex mode for readying
binary data.

On the SARA-R4 this mode limits the receiving MTU to 512 which is
unacceptable.

Signed-off-by: Michael Scott <mike@foundries.io>

mike-scott committed:
drivers: modem: sara-r4: use +USO[RF|RD] based on proto

Regardless of whether from is set, let's use the correct socket read
commands for UDP and TCP.

Signed-off-by: Michael Scott <mike@foundries.io>

mike-scott commented:
Rebased on today's master with the last set of changes:
- Remove hex mode.  It limits the send/receive MTU on SARA-R4 to 512 which is too restrictive.  We can now handle this inside the modem commands by returning EAGAIN if the command doesn't have enough data to finish parsing.
- parse_params() shouldn't parse as many delimiters as it can, instead it should honor the arg_count value and stop once it receives that # of params
- And a few minor changes

mike-scott commented:
> I tested this PR with Sara R410, with following results:
> 
> * MQTT client works fine
> * echo_client (TCP) works fine with small chunks (up to 1023 bytes)
> * echo_client (TCP) fails with chunks >= 1024 bytes
> 
> I then discovered that in case of these failures, the +USOWR fails with an error message from the modem ("ERROR"), immediately followed by a +UUSORD message from the modem.
> 
> To me it looks like a solution could be to let the application poll for incoming data in between multiple calls to send(). So should this be solved in the driver, or rather in the application?

I have a separate PR to fix echo_client for net interfaces where the MTU is < 1280: https://github.com/zephyrproject-rtos/zephyr/pull/22448

mike-scott commented:
> run MQTT publish loop. I invest

Please retry with the latest version of this PR.  I had a few more fixes to land on top today.  Sorry about that.

gpaquet85 commented:
> 
> 
> > @mike-scott I made some tries. For the moment basic MQTT publish is ok. Thanks for this. However I still have some fails when I run MQTT publish loop. I investigate right now on this and let you know.
> 
> Please retry with the latest version of this PR. I had a few more fixes to land on top today. Sorry about that.

@mike-scott I made some tries. It works. Really thanks for your support on this case :)

rlubos commented:
@mike-scott Just out of curiosity, given the offloading rework got merged sooner than I expected, you had to rebase on top of this work. Does the new approach work fine for you, did you encounter any issues with these changes? (apart from extra work with conflict solving I guess)

mike-scott commented:
Rebased on today's master (post TOPIC-GPIO merge):
- added patches to fix build issues due to GPIO changes

mike-scott commented:
@jukkar Does this PR have any impact on the new GSM driver?  I don't think it does..

jukkar commented:
> @mike-scott  Does this PR have any impact on the new GSM driver? I don't think it does..

Looks quite clean to me so there should be no issues with GSM driver.

jukkar commented:
@mike-scott Some of these features would be useful to have in GSM driver too if the modem is a SARA modem, for example "support for network indication on SARA modems". Currently we need to add this to GSM modem init function, but I wonder if we could isolate these generic and common settings to some library or similar, which could be used from both the GSM driver and offloading driver.

jukkar commented on drivers/modem/Kconfig.ublox-sara-r4:
```diff
@@ -77,4 +77,21 @@ config MODEM_UBLOX_SARA_R4_INIT_PRIORITY
+
+config MODEM_UBLOX_SARA_R4_NETWORK_STATUS_PIN
+	string "Network status indication gpio_id"
+	default "16"
```
Why this is a string and not an int?

mike-scott replied:
Woops.  I picked up this patch from @weinholtendian -- I will make this an int.

jukkar commented on drivers/modem/modem_cmd_handler.c:
```diff
@@ -266,13 +275,14 @@ static void cmd_handler_process(struct modem_cmd_handler *cmd_handler,
+					      data->match_buf_len - 1,
 					      data->rx_buf, 0, len);
-		if (data->match_buf_len < match_len) {
+		if (data->match_buf_len - 1 < match_len) {
```
Not really an issue but it would be nice to the reader if () would be used here, so `if ((data->match_buf_len - 1) < match_len) {`

mike-scott replied:
I'll add parens.

jukkar commented on drivers/modem/ublox-sara-r4.c:
```diff
@@ -1205,13 +1276,97 @@ static bool offload_is_supported(int family, int type, int proto)
+	}
+
+	if (service) {
+		port = strtol(service, NULL, 10);
```
Could we call `modem_atoi()` here so that we can handle error cases properly.

mike-scott replied:
Yep.  Will change this.

mike-scott commented:
Need to rebase this again (I'll remove the last 2 patches with fixes that have been merged).   And address @jukkar comments.

mike-scott commented:
> @mike-scott Just out of curiosity, given the offloading rework got merged sooner than I expected, you had to rebase on top of this work. Does the new approach work fine for you, did you encounter any issues with these changes? (apart from extra work with conflict solving I guess)

No, so far so good.  @rlubos thanks for the rework :)

mike-scott commented on drivers/modem/ublox-sara-r4.c:
```diff
@@ -1068,23 +1102,44 @@ static ssize_t offload_recvfrom(void *obj, void *buf, size_t len,
+	 * Binary and ASCII mode allows sending 1024 bytes to the socket
+	 * in one command
+	 */
+	if (next_packet_size > 1024) {
```
Hah, yes.  Good catch.  I have the define already and didn't use it here.

mike-scott commented on drivers/modem/ublox-sara-r4.c:
```diff
@@ -1205,13 +1276,97 @@ static bool offload_is_supported(int family, int type, int proto)
+	/* AT+UDNSRN=#,"[+128]"\r+NUL == 17 + 128 == 145
+	 * (allows for domain name of up to 128 characters)
+	 */
+	char sendbuf[145];
```
Sounds good.  I'll rework this.

mike-scott commented:
- Rebased again on today's master
- Addressed @jukkar comments

NOTE: I also shortened `MODEM_UBLOX_SARA_R4_NETWORK_STATUS` config to:
`MODEM_UBLOX_SARA_R4_NET_STATUS`.

To help with review, here's the diff from the previous version:
```
diff --git a/drivers/modem/Kconfig.ublox-sara-r4 b/drivers/modem/Kconfig.ublox-sara-r4
index 17736e89eb..c6a373d6eb 100644
--- a/drivers/modem/Kconfig.ublox-sara-r4
+++ b/drivers/modem/Kconfig.ublox-sara-r4
@@ -77,21 +77,21 @@ config MODEM_UBLOX_SARA_R4_INIT_PRIORITY
 	  Note that the priority needs to be lower than the net stack
 	  so that it can start before the networking sub-system.
 
-config MODEM_UBLOX_SARA_R4_NETWORK_STATUS
+config MODEM_UBLOX_SARA_R4_NET_STATUS
 	bool "Enable support for network status indication"
 	help
 	  Choose this setting to use a modem GPIO pin as network indication.
 
-if MODEM_UBLOX_SARA_R4_NETWORK_STATUS
+if MODEM_UBLOX_SARA_R4_NET_STATUS
 
-config MODEM_UBLOX_SARA_R4_NETWORK_STATUS_PIN
-	string "Network status indication gpio_id"
-	default "16"
+config MODEM_UBLOX_SARA_R4_NET_STATUS_PIN
+	int "Network status indication GPIO ID"
+	default 16
 	help
 	  This setting is used to configure one of the modem's GPIO pins
 	  as a network status indication.  See the manual for the gpio ids
 	  and how they map to pin numbers.
 
-endif # MODEM_UBLOX_SARA_R4_NETWORK_STATUS
+endif # MODEM_UBLOX_SARA_R4_NET_STATUS
 
 endif # MODEM_UBLOX_SARA
diff --git a/drivers/modem/modem_cmd_handler.c b/drivers/modem/modem_cmd_handler.c
index 4d795d54ee..fcab704510 100644
--- a/drivers/modem/modem_cmd_handler.c
+++ b/drivers/modem/modem_cmd_handler.c
@@ -279,7 +279,7 @@ static void cmd_handler_process(struct modem_cmd_handler *cmd_handler,
 		match_len = net_buf_linearize(data->match_buf,
 					      data->match_buf_len - 1,
 					      data->rx_buf, 0, len);
-		if (data->match_buf_len - 1 < match_len) {
+		if ((data->match_buf_len - 1) < match_len) {
 			LOG_ERR("Match buffer size (%zu) is too small for "
 				"incoming command size: %u!  Truncating!",
 				data->match_buf_len - 1, match_len);
diff --git a/drivers/modem/ublox-sara-r4.c b/drivers/modem/ublox-sara-r4.c
index 5241bc018d..1bb846b778 100644
--- a/drivers/modem/ublox-sara-r4.c
+++ b/drivers/modem/ublox-sara-r4.c
@@ -206,11 +206,11 @@ static ssize_t send_socket_data(struct modem_socket *sock,
 	}
 
 	/*
-	 * Binary and ASCII mode allows sending 1024 bytes to the socket
-	 * in one command
+	 * Binary and ASCII mode allows sending MDM_MAX_DATA_LENGTH bytes to
+	 * the socket in one command
 	 */
-	if (buf_len > 1024) {
-		buf_len = 1024;
+	if (buf_len > MDM_MAX_DATA_LENGTH) {
+		buf_len = MDM_MAX_DATA_LENGTH;
 	}
 
 	/* The number of bytes written will be reported by the modem */
@@ -729,11 +729,11 @@ static void modem_reset(void)
 		/* use external SIM */
 		SETUP_CMD_NOHANDLE("AT+UGPIOC=23,0,0"),
 #endif
-#if defined(CONFIG_MODEM_UBLOX_SARA_R4_NETWORK_STATUS_PIN)
+#if defined(CONFIG_MODEM_UBLOX_SARA_R4_NET_STATUS_PIN)
 		/* enable the network status indication */
 		SETUP_CMD_NOHANDLE("AT+UGPIOC="
-				   CONFIG_MODEM_UBLOX_SARA_R4_NETWORK_STATUS_PIN
-				   ",2"),
+			STRINGIFY(CONFIG_MODEM_UBLOX_SARA_R4_NET_STATUS_PIN)
+			",2"),
 #endif
 		/* UNC messages for registration */
 		SETUP_CMD_NOHANDLE("AT+CREG=1"),
@@ -1130,11 +1130,11 @@ static ssize_t offload_recvfrom(void *obj, void *buf, size_t len,
 	}
 
 	/*
-	 * Binary and ASCII mode allows sending 1024 bytes to the socket
-	 * in one command
+	 * Binary and ASCII mode allows sending MDM_MAX_DATA_LENGTH bytes to
+	 * the socket in one command
 	 */
-	if (next_packet_size > 1024) {
-		next_packet_size = 1024;
+	if (next_packet_size > MDM_MAX_DATA_LENGTH) {
+		next_packet_size = MDM_MAX_DATA_LENGTH;
 	}
 
 	snprintk(sendbuf, sizeof(sendbuf), "AT+USO%s=%d,%d",
@@ -1288,10 +1288,8 @@ static int offload_getaddrinfo(const char *node, const char *service,
 	struct modem_cmd cmd = MODEM_CMD("+UDNSRN: ", on_cmd_dns, 1U, ",");
 	u32_t port = 0U;
 	int ret;
-	/* AT+UDNSRN=#,"[+128]"\r+NUL == 17 + 128 == 145
-	 * (allows for domain name of up to 128 characters)
-	 */
-	char sendbuf[145];
+	/* DNS command + 128 bytes for domain name parameter */
+	char sendbuf[sizeof("AT+UDNSRN=#,'[]'\r") + 128];
 
 	/* init result */
 	(void)memset(&result, 0, sizeof(result));
@@ -1318,7 +1316,7 @@ static int offload_getaddrinfo(const char *node, const char *service,
 	}
 
 	if (service) {
-		port = strtol(service, NULL, 10);
+		port = ATOI(service, 0U, "port");
 		if (port < 1 || port > USHRT_MAX) {
 			return EAI_SERVICE;
 		}
```

jukkar reviewed (approved):
LGTM

gpaquet85 commented on drivers/modem/ublox-sara-r4.c:
```diff
@@ -647,9 +640,17 @@ static int pin_init(void)
 	LOG_DBG("Waiting for MDM_VINT_PIN = 0");
 
-	do {
+	while (modem_pin_read(&mctx, MDM_VINT) != MDM_VINT_DISABLE) {
```
what do you think about adding timeout test in order to be sure to exit loop ?
I had one case where I was always blocked in this loop.

mike-scott replied:
@gpaquet85 Can we hold off on that until after this is merged?  The loop doesn't change here, other than the values are checked at the top of the loop.  We can address it in a new PR with discussion.

gpaquet85 replied:
@mike-scott ok no problem. Indeed it can be adressed later

erwango reviewed (approved):
Approved (shield-wise)

mike-scott commented:
@gpaquet85 @hwilmers @WilliamGFish @weinholtendian Please weigh in as to whether this is ready to be merged.   I'd like those that are interested to have a voice.

WilliamGFish commented:
I'll have a look next week. Billy..On 7 Feb 2020 20:15, Michael Scott <notifications@github.com> wrote:@gpaquet85 @hwilmers @WilliamGFish @weinholtendian Please weigh in as to whether this is ready to be merged.   I'd like those that are interested to have a voice.

—You are receiving this because you were mentioned.Reply to this email directly, view it on GitHub, or unsubscribe.

mike-scott commented:
@jukkar I'm hoping this PR is classified as a "bugfix" and is still acceptable for v2.2?

WilliamGFish commented:
Was all looking good until I tried to use the U2 version then an odd **_CME ERROR: 133 | Requested service option not subscribed_**  

Turns out a typo which I think came from a cut-n-paste from an order version(_ublox-sara-r4.c_):
```
#if defined(CONFIG_MODEM_UBLOX_SARA_U2)
		/* set the APN */
		SETUP_CMD_NOHANDLE("AT+UPSD=0,1,\""
				CONFIG_MODEM_UBLOX_SARA_R4_MANUAL_MCCMNO "\""),

```
Should be:
```
#if defined(CONFIG_MODEM_UBLOX_SARA_U2)
		/* set the APN */
		SETUP_CMD_NOHANDLE("AT+UPSD=0,1,\""
				CONFIG_MODEM_UBLOX_SARA_R4_APN "\""),

```

mike-scott commented:
@WilliamGFish Ok, easy fix.  Let me put a PR up.  Can you open an issue for that?  Zephyr is in bug-fix only mode now.

WilliamGFish commented:
> @WilliamGFish Ok, easy fix. Let me put a PR up. Can you open an issue for that? Zephyr is in bug-fix only mode now.

Issue raised
driver: modem: sara-u2 error when connecting #22689

Brietoe commented:
I've a question, pardon my ignorance if I'm overlooking something here.

I've implemented this driver and have it working for the most part. When I open a socket and send data to an echo server, I get back a response which shows as an unsolicited message with a 'number bytes ready' count. When I use the recv() function to read incoming data, I'm unable to get data that arrived before I called the recv() function but has already been handled with the modem_socket_data_ready() function. It looks like, while in modem_socket_wait_data() it only waits for fresh data to come in and doesn't detect recently buffered data.

Is this the proper functionality of recv() or am I doing something wrong?

weinholtendian commented:
> Is this the proper functionality of recv() or am I doing something wrong?

@Brietoe TCP data should never disappear merely because recv() was called too late. The driver is in control of when it reads from the modem's socket buffer, and it does not need to read if it doesn't have anywhere to save the data. Sounds to me like a pretty severe bug if data disappears this way. UDP is a slightly different story.

Brietoe commented:
@weinholtendian Okay, that's what I thought. I don't want to jump to conclusions and claim that the offload_recv() does indeed suffer from this issue because I had to make some modifications to the code to fit my use case. The data is still accessible through workarounds but I'm just not able to retrieve it using recv().

I'd love to know if anyone else is also experiencing this issue.

WilliamGFish commented:
> Other than TLS support and possibly DNS, what functionality were you looking for on the NET_OFFLOAD APIs?

Hey @mike-scott 
Looking at the particle boards and TLS is now a big question. I'm considering getting a PPP set-up option for Boron if I can.

Billy..
